compiler/noirc_evaluator/src/ssa/ir/types.rs

```diff
@@ -31,12 +31,31 @@ impl NumericType {
 
     /// Returns true if the given Field value is within the numeric limits
     /// for the current NumericType.
-    pub(crate) fn value_is_within_limits(self, field: FieldElement) -> bool {
+    ///
+    /// Note that if `negative` is true, it's expected that the field value holds the negative value
+    /// already casted to the range of `self` (for example, `-1` for a `i8` or `u8` type is
+    /// should be passed as `255`).
+    pub(crate) fn value_is_within_limits(self, field: FieldElement, negative: bool) -> bool {
         match self {
-            NumericType::Signed { bit_size } | NumericType::Unsigned { bit_size } => {
+            NumericType::Unsigned { bit_size } => {
+                if negative {
+                    return false;
+                }
                 let max = 2u128.pow(bit_size) - 1;
                 field <= max.into()
             }
+            NumericType::Signed { bit_size } => {
+                if negative {
+                    let base = 1_u128 << bit_size;
+                    // Recover the original value (a similar operation is done in compiler/noirc_frontend/src/monomorphization/mod.rs)
+                    let field = FieldElement::from(base) - field;
```

Comment:
I'm not happy about putting this reverse logic here... maybe `value_is_within_limits` should be called before the logic on the monomorphization side is done... but I didn't want to change things too much.